source_files/watchdog.c

```diff
@@ -79,53 +88,107 @@ void *watchdog_timer(watchdog_arg_t *arg)
     sigaddset(&signal_mask, SIGTERM);
     pthread_sigmask(SIG_BLOCK, &signal_mask, NULL);
 
-    period = arg->period;
-    load = arg->load;
+    signal(SIGALRM, sigalrm_handler);
+
+    period = arg->period*1000;
+    load = arg->load*1000;
     idle = period - load;
     timeout = arg->timeout;
     loadvar = arg->loadvar;
 
+    watchdog_thread = pthread_self();
 
     clock_gettime(CLOCK_REALTIME, &start_ts);
     time = 0;
     /* TODO: I don't like that the control flow depends on the period variable,
      * it is not wrong but confusing
      */
+
     while(period > 0){
         // cycles++;
 
         clock_gettime(CLOCK_REALTIME, &current);
-        advance = ((current.tv_sec - start_ts.tv_sec) * 1000000 + (current.tv_nsec - start_ts.tv_nsec) / 1000) % period;
+        advance = ((current.tv_sec - start_ts.tv_sec) * 1000000000 + (current.tv_nsec - start_ts.tv_nsec) ) % period;
         load_reduction = (load * advance) / period;
         idle_reduction = advance - load_reduction;
 
+        target.tv_nsec = (load - load_reduction) % 1000000000;
+        target.tv_sec  = (load - load_reduction) / 1000000000;
 #ifdef ENABLE_VTRACING
         VT_USER_START("WD_HIGH");
 #endif
 #ifdef ENABLE_SCOREP
         SCOREP_USER_REGION_BY_NAME_BEGIN("WD_HIGH", SCOREP_USER_REGION_TYPE_COMMON);
 #endif
-
-        sleepret = usleep(load - load_reduction);
-        while(sleepret != 0){ /* sometimes usleep fails, this is to be very sure it works */
-            sleepret = usleep(load - load_reduction);
+        sleepret = nanosleep(&target,&remaining);
```

Comment:
These 20 lines are redundant, and you should move them to a function (or macro, if you're worried about inlining).